### ISSUE TITLE
Investigate Creation of Empty Value Providers in TechEmpower scenarios

### ISSUE THREAD

**Issue by rynowak** [aspnet/Mvc#4368]
Sampling data from HelloWorldMvc

![image](https://cloud.githubusercontent.com/assets/1430011/14116047/be43710e-f592-11e5-8257-990d70588237.png)

Creating this value providers might be too expensive when there's no data

**Comment by mnltejaswini** [comment]
The cost of creating  Form and JQueryForm value providers is from the check Request.HasFormContentType . The check involves the following hot spots
![image](https://cloud.githubusercontent.com/assets/8011073/14468240/b1bc73a2-0093-11e6-9f5d-d58ec42961c2.png)

![image](https://cloud.githubusercontent.com/assets/8011073/14468253/bf65ed26-0093-11e6-9f69-5d0af87318e6.png)

Tried few things in Kestrel and HttpAbstractions repo. 
1. Currently https://github.com/aspnet/KestrelHttpServer/blob/dev/src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.Generated.cs

FastFeatureGet(Type key) the checks are made as 
```
if (key == typeof(global::Microsoft.AspNetCore.Http.Features.IHttpRequestFeature))
            {
                return _currentIHttpRequestFeature;
            }
```
Instead of using typeof, is there any benefit of using the static Type fields in the same class. But benchmark tests corresponding to them didn't show any improvement

2. Currently https://github.com/aspnet/KestrelHttpServer/blob/dev/src/Microsoft.AspNetCore.Server.Kestrel/Http/FrameHeaders.Generated.cs 
TryGetValueFast(string key, out StringValues value) , the header keys are compared for equality using string.Equals(). Tried with alternative approaches like string.Compare with no luck.

The current existing code in HttpAbstractions is optimized to maximum extent. 

3. Another approach is in MVC repo to lazy create these value providers . This would be an ideal solution. But this has two caveats
1. Currently value providers are a property on the Controller context. Creation of value providers is async and this design of value providers as property doesn't work
2. We can change the property as a method but this design doesn't fit in the current code base.

Hence there are no simple and straightforward optimizations to handle this issue other than design changes in MVC repo. For that reason I am holding onto this issue for now .

**Comment by rynowak** [comment]
Fixed by 4d63ffa879370f6945d011e0d1e2a5f55bbac119 - value providers are lazy and don't get created unless you have parameters/modelbinding
